# 2. Weather model

In [6]:
import statsmodels.api as sm
import matplotlib.pyplot as plt
import seaborn as sns
import numpy as np
import pandas  as pd

In [42]:
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [8]:
weather_df["Temp_Difference"] = weather_df["Temperature (C)"]-weather_df["Apparent Temperature (C)"]

In [22]:
X = weather_df[['Humidity', 'Wind Speed (km/h)']]
y = weather_df["Temperature (C)"]

In [23]:
X_cons = sm.add_constant(X)
results = sm.OLS(y, X_cons).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Temperature (C)   R-squared:                       0.418
Model:                            OLS   Adj. R-squared:                  0.418
Method:                 Least Squares   F-statistic:                 3.470e+04
Date:                Thu, 17 Feb 2022   Prob (F-statistic):               0.00
Time:                        13:59:38   Log-Likelihood:            -3.2838e+05
No. Observations:               96453   AIC:                         6.568e+05
Df Residuals:                   96450   BIC:                         6.568e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                37.8674      0.108    350.251      0.000      37.656      38.079
Humidity            -32.4374      0.123   -263.415      0.000     -32.679     -32.196
Wind Speed (km/h)    -0.1939      0.003    -55.701      0.000      -0.201      -0.187
==============================================================================
Omnibus:                     3415.546   Durbin-Watson:                   0.054
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3844.707
Skew:                          -0.458   Prob(JB):                         0.00
Kurtosis:                       3.344   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [11]:
X = weather_df[['Humidity', 'Wind Speed (km/h)']]
y = weather_df["Temp_Difference"]
X_cons = sm.add_constant(X)
results = sm.OLS(y, X_cons).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Temp_Difference   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Thu, 17 Feb 2022   Prob (F-statistic):               0.00
Time:                        13:47:30   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**When we use temperature as a target variable, the R square value is  around 0.418. On the other hand, when we use temperature difference as a target variable, the R squared value is around 0.288. So for both model, R-Squared values are low. but in the temperature difference target variable is too low.**

In [16]:
weather_df["interaction_humid_wind"] = weather_df['Humidity']*weather_df['Wind Speed (km/h)']
X = weather_df[['Humidity', 'Wind Speed (km/h)',"interaction_humid_wind"]]
y = weather_df["Temperature (C)"]
X_cons = sm.add_constant(X)
results = sm.OLS(y, X_cons).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Temperature (C)   R-squared:                       0.419
Model:                            OLS   Adj. R-squared:                  0.419
Method:                 Least Squares   F-statistic:                 2.315e+04
Date:                Thu, 17 Feb 2022   Prob (F-statistic):               0.00
Time:                        13:51:46   Log-Likelihood:            -3.2837e+05
No. Observations:               96453   AIC:                         6.567e+05
Df Residuals:                   96449   BIC:                         6.568e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     37.9469      0.109    348.232      0.000      37.733      38.161
Humidity                 -33.2946      0.193   -172.644      0.000     -33.673     -32.917
Wind Speed (km/h)         -0.2505      0.010    -24.097      0.000      -0.271      -0.230
interaction_humid_wind     0.4460      0.077      5.775      0.000       0.295       0.597
==============================================================================
Omnibus:                     3341.982   Durbin-Watson:                   0.054
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3746.828
Skew:                          -0.454   Prob(JB):                         0.00
Kurtosis:                       3.329   Cond. No.                         121.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

**When we add the interaction, the R-Squared values is increasing a little bit from 0.418 to 0.419**

In [17]:
X = weather_df[['Humidity', 'Wind Speed (km/h)',"Visibility (km)"]]
y = weather_df["Temperature (C)"]
X_cons = sm.add_constant(X)
results = sm.OLS(y, X_cons).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Temperature (C)   R-squared:                       0.449
Model:                            OLS   Adj. R-squared:                  0.449
Method:                 Least Squares   F-statistic:                 2.618e+04
Date:                Thu, 17 Feb 2022   Prob (F-statistic):               0.00
Time:                        13:53:41   Log-Likelihood:            -3.2580e+05
No. Observations:               96453   AIC:                         6.516e+05
Df Residuals:                   96449   BIC:                         6.516e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                31.0402      0.141    220.313      0.000      30.764      31.316
Humidity            -29.0927      0.128   -226.641      0.000     -29.344     -28.841
Wind Speed (km/h)    -0.1988      0.003    -58.629      0.000      -0.205      -0.192
Visibility (km)       0.4273      0.006     72.896      0.000       0.416       0.439
==============================================================================
Omnibus:                     3490.262   Durbin-Watson:                   0.072
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3926.930
Skew:                          -0.466   Prob(JB):                         0.00
Kurtosis:                       3.331   Cond. No.                         131.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [18]:
weather_df["interaction_humid_wind"] = weather_df['Humidity']*weather_df['Wind Speed (km/h)']
X = weather_df[['Humidity', 'Wind Speed (km/h)',"interaction_humid_wind","Visibility (km)"]]
y = weather_df["Temperature (C)"]
X_cons = sm.add_constant(X)
results = sm.OLS(y, X_cons).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Temperature (C)   R-squared:                       0.449
Model:                            OLS   Adj. R-squared:                  0.449
Method:                 Least Squares   F-statistic:                 1.963e+04
Date:                Thu, 17 Feb 2022   Prob (F-statistic):               0.00
Time:                        13:55:30   Log-Likelihood:            -3.2580e+05
No. Observations:               96453   AIC:                         6.516e+05
Df Residuals:                   96448   BIC:                         6.517e+05
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
const                     31.1120      0.205    151.930      0.000      30.711      31.513
Humidity                 -29.1866      0.233   -125.372      0.000     -29.643     -28.730
Wind Speed (km/h)         -0.2048      0.013    -15.932      0.000      -0.230      -0.180
interaction_humid_wind     0.0085      0.018      0.483      0.629      -0.026       0.043
Visibility (km)            0.4270      0.006     72.482      0.000       0.415       0.439
==============================================================================
Omnibus:                     3489.948   Durbin-Watson:                   0.072
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             3926.966
Skew:                          -0.466   Prob(JB):                         0.00
Kurtosis:                       3.332   Cond. No.                         246.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

- **The both of them R-squared and adj. R-squared values increase from 0.418 to 0.449.**
- **And when we add the interaction variable nothing change in term of R-Squared values**

Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

**Actually lowest AIC and BIC values are taken from the model that uses the temperature difference as a target variable**

# 2. House prices model

In [46]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [47]:
miss_val_ind = (house_prices_df.isnull().sum().sort_values(ascending=False)/len(house_prices_df)*100).head(10).index
house_prices_df.drop(miss_val_ind[0:5],axis=1,inplace=True)
house_prices_df["LotFrontage"].fillna(house_prices_df["LotFrontage"].mean(),inplace=True)
house_prices_df.dropna(inplace=True)
house_prices_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1338 entries, 0 to 1459
Data columns (total 76 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   Id             1338 non-null   int64  
 1   MSSubClass     1338 non-null   int64  
 2   MSZoning       1338 non-null   object 
 3   LotFrontage    1338 non-null   float64
 4   LotArea        1338 non-null   int64  
 5   Street         1338 non-null   object 
 6   LotShape       1338 non-null   object 
 7   LandContour    1338 non-null   object 
 8   Utilities      1338 non-null   object 
 9   LotConfig      1338 non-null   object 
 10  LandSlope      1338 non-null   object 
 11  Neighborhood   1338 non-null   object 
 12  Condition1     1338 non-null   object 
 13  Condition2     1338 non-null   object 
 14  BldgType       1338 non-null   object 
 15  HouseStyle     1338 non-null   object 
 16  OverallQual    1338 non-null   int64  
 17  OverallCond    1338 non-null   int64  
 18  YearBuil

In [48]:
obj_ind = house_prices_df.dtypes[house_prices_df.dtypes=="O"].index
for i in obj_ind:
    if house_prices_df[i].nunique()>5:
        house_prices_df.drop(i,axis=1,inplace=True)

In [49]:
obj_ind = house_prices_df.dtypes[house_prices_df.dtypes=="O"]
corr_ind = house_prices_df.corr()["SalePrice"].sort_values(ascending=False).index
house_prices_df.drop(corr_ind[11:],axis=1,inplace=True)
complet_df = pd.get_dummies(house_prices_df,drop_first=True)

In [50]:
X = complet_df.drop("SalePrice",axis=1)
y = complet_df["SalePrice"]
X_cons = sm.add_constant(X)
results = sm.OLS(y, X_cons).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.833
Model:                            OLS   Adj. R-squared:                  0.822
Method:                 Least Squares   F-statistic:                     78.25
Date:                Thu, 17 Feb 2022   Prob (F-statistic):               0.00
Time:                        15:09:01   Log-Likelihood:                -15789.
No. Observations:                1338   AIC:                         3.174e+04
Df Residuals:                    1257   BIC:                         3.216e+04
Df Model:                          80                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const              -2.574e+05   2.03e+05     -1.270      0.204   -6.55e+05     1.4e+05
OverallQual         1.208e+04   1349.710      8.953      0.000    9435.984    1.47e+04
YearBuilt             11.2596     83.787      0.134      0.893    -153.117     175.636
YearRemodAdd         126.4487     74.345      1.701      0.089     -19.405     272.303
TotalBsmtSF           11.2201      5.889      1.905      0.057      -0.334      22.774
1stFlrSF               4.6375      6.229      0.745      0.457      -7.582      16.857
GrLivArea             49.0804      4.202     11.680      0.000      40.836      57.325
FullBath            -841.9030   2716.185     -0.310      0.757   -6170.658    4486.852
TotRmsAbvGrd         368.0245   1175.104      0.313      0.754   -1937.357    2673.406
GarageCars          1.482e+04   3058.178      4.846      0.000    8821.585    2.08e+04
GarageArea            -5.7695     10.002     -0.577      0.564     -25.391      13.852
MSZoning_FV         3.998e+04   1.43e+04      2.797      0.005    1.19e+04     6.8e+04
MSZoning_RH         3.597e+04   1.67e+04      2.156      0.031    3243.449    6.87e+04
MSZoning_RL         3.369e+04   1.35e+04      2.499      0.013    7241.845    6.01e+04
MSZoning_RM         2.337e+04   1.35e+04      1.735      0.083   -3052.616    4.98e+04
Street_Pave         2158.7786   1.86e+04      0.116      0.908   -3.44e+04    3.87e+04
LotShape_IR2        8848.9099   5853.643      1.512      0.131   -2635.078    2.03e+04
LotShape_IR3       -4.141e+04   1.16e+04     -3.583      0.000   -6.41e+04   -1.87e+04
LotShape_Reg       -2088.2481   2231.901     -0.936      0.350   -6466.911    2290.414
LandContour_HLS     2.588e+04   7213.954      3.587      0.000    1.17e+04       4e+04
LandContour_Low     2.439e+04   8817.345      2.767      0.006    7096.145    4.17e+04
LandContour_Lvl     1.897e+04   5377.199      3.527      0.000    8418.194    2.95e+04
Utilities_NoSeWa   -3.344e+04   3.42e+04     -0.979      0.328      -1e+05    3.36e+04
LotConfig_CulDSac   1.214e+04   4429.040      2.742      0.006    3455.226    2.08e+04
LotConfig_FR2      -8557.9699   5680.357     -1.507      0.132   -1.97e+04    2586.056
LotConfig_FR3      -4154.6263   1.72e+04     -0.241      0.809    -3.8e+04    2.96e+04
LotConfig_Inside     798.4832   2498.025      0.320      0.749   -4102.274    5699.241
LandSlope_Mod       9081.2389   5570.908      1.630      0.103   -1848.064       2e+04
LandSlope_Sev       3878.1400   1.16e+04      0.336      0.737   -1.88e+04    2.66e+04
BldgType_2fmCon    -1.439e+04   8126.180     -1.770      0.077   -3.03e+04    1555.529
BldgType_Duplex      -3.5e+04   7129.585     -4.909      0.000    -4.9e+04    -2.1e+04
BldgType_Twnhs     -2.056e+04   6258.670     -3.284      0.001   -3.28e+04   -8277.717
BldgType_TwnhsE    -1.511e+04   3924.4

Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.

- **The R squared value is around 0.833. we can say that the model is good to use**
- **F-test is around 78.25**
- **AIC and BIC values around 3.174e+04 and 3.216e+04.**

Do you think your model is satisfactory? If so, why?

**Actually, we can said that consider the R-Squared value, the model is good**

In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables.

In [51]:
drop_data = ["GarageCars","MSZoning","Street","LotShape","LandContour","Utilities","LotConfig","LotConfig","LandSlope","BldgType","ExterCond","Foundation","BsmtCond","BsmtExposure","Heating","HeatingQC","Electrical","GarageFinish","GarageQual","PavedDrive"]
house_prices_df.drop(drop_data,axis=1,inplace=True)
obj_ind = house_prices_df.dtypes[house_prices_df.dtypes=="O"]
corr_ind = house_prices_df.corr()["SalePrice"].sort_values(ascending=False).index
house_prices_df.drop(corr_ind[11:],axis=1,inplace=True)
complet_df = pd.get_dummies(house_prices_df,drop_first=True)
X = complet_df.drop("SalePrice",axis=1)
y = complet_df["SalePrice"]
X_cons = sm.add_constant(X)
results = sm.OLS(y, X_cons).fit()
results.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.795
Model:                            OLS   Adj. R-squared:                  0.791
Method:                 Least Squares   F-statistic:                     195.4
Date:                Thu, 17 Feb 2022   Prob (F-statistic):               0.00
Time:                        15:09:04   Log-Likelihood:                -15926.
No. Observations:                1338   AIC:                         3.191e+04
Df Residuals:                    1311   BIC:                         3.205e+04
Df Model:                          26                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const              -5.895e+05    1.8e+05     -3.280      0.001   -9.42e+05   -2.37e+05
OverallQual         1.427e+04   1373.877     10.387      0.000    1.16e+04     1.7e+04
YearBuilt            179.5368     66.511      2.699      0.007      49.056     310.017
YearRemodAdd         140.7579     74.732      1.884      0.060      -5.849     287.365
TotalBsmtSF            7.9262      6.085      1.302      0.193      -4.012      19.865
1stFlrSF              14.7933      6.379      2.319      0.021       2.280      27.307
GrLivArea             50.5018      4.288     11.777      0.000      42.089      58.914
FullBath           -3316.8274   2799.083     -1.185      0.236   -8807.998    2174.343
TotRmsAbvGrd         424.9901   1156.646      0.367      0.713   -1844.089    2694.069
GarageArea            35.4786      7.107      4.992      0.000      21.537      49.420
MasVnrType_BrkFace  8455.2565   9596.588      0.881      0.378   -1.04e+04    2.73e+04
MasVnrType_None     9114.7430   9503.821      0.959      0.338   -9529.616    2.78e+04
MasVnrType_Stone    9218.0022   1.02e+04      0.907      0.365   -1.07e+04    2.92e+04
ExterQual_Fa       -3.708e+04   1.61e+04     -2.304      0.021   -6.87e+04   -5502.064
ExterQual_Gd       -1.735e+04   6735.910     -2.576      0.010   -3.06e+04   -4135.713
ExterQual_TA       -2.335e+04   7510.647     -3.109      0.002   -3.81e+04   -8614.463
BsmtQual_Fa        -4.187e+04   9018.519     -4.642      0.000   -5.96e+04   -2.42e+04
BsmtQual_Gd        -3.266e+04   4614.135     -7.079      0.000   -4.17e+04   -2.36e+04
BsmtQual_TA        -3.711e+04   5617.484     -6.607      0.000   -4.81e+04   -2.61e+04
CentralAir_Y        1.676e+04   5423.145      3.090      0.002    6121.104    2.74e+04
KitchenQual_Fa     -3.164e+04   9783.061     -3.234      0.001   -5.08e+04   -1.24e+04
KitchenQual_Gd     -2.944e+04   5072.272     -5.805      0.000   -3.94e+04   -1.95e+04
KitchenQual_TA      -3.61e+04   5767.283     -6.260      0.000   -4.74e+04   -2.48e+04
GarageCond_Fa      -1.231e+04   2.66e+04     -0.463      0.643   -6.44e+04    3.98e+04
GarageCond_Gd      -2470.1453   2.84e+04     -0.087      0.931   -5.82e+04    5.33e+04
GarageCond_Po      -2.171e+04   2.95e+04     -0.736      0.462   -7.96e+04    3.61e+04
GarageCond_TA      -7008.6610   2.57e+04     -0.272      0.785   -5.75e+04    4.35e+04
==============================================================================
Omnibus:                      785.940   Durbin-Watson:                   1.985
Prob(Omnibus):                  0.000   Jarque-Bera (JB):           109198.329
Skew:                          -1.715   Prob(JB):                         0.00
Kurtosis:                      47.124   Cond. No.                     6.63e+05
==============================================================================

For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

- **When we remove some of features, the R-Squared value decrease but AIC and BIC values decrease a little bit. Moreover, the F-test score increase to much.**
- **I think the second model is better. R-Square value of first is better but when we consider the other parameter, in the end, i decided that second model is better**